src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.cc

```diff
@@ -0,0 +1,360 @@
+    std::cout << "Unknown domain in pipe drain evaluator" << std::endl;
+    AMANZI_ASSERT(0);
+  }
+}
+
+
+Teuchos::RCP<Evaluator>
+PipeDrainEvaluator::Clone() const
+{
+  return Teuchos::rcp(new PipeDrainEvaluator(*this));
+}
+
+
+void
+PipeDrainEvaluator::CreateCellMap(const State& S)
+{
+  // grab the meshes
+  Teuchos::RCP<const AmanziMesh::Mesh> pipe_mesh = S.GetMesh(pipe_domain_name_);
+  Teuchos::RCP<const AmanziMesh::Mesh> surface_mesh = S.GetMesh(sw_domain_name_);
+
+  // get the manhole locations/ cell maps
+  Tag tag = my_keys_.front().second;
+
+  const Epetra_MultiVector& mnhMask =
+    *S.GetPtr<CompositeVector>(mask_key_, tag)->ViewComponent("cell", false);
+
+  // loop over mesh cells and create map
+  int ncells_pipe =
+    pipe_mesh->getNumEntities(AmanziMesh::Entity_kind::CELL, AmanziMesh::Parallel_kind::OWNED);
+  int ncells_sw =
+    surface_mesh->getNumEntities(AmanziMesh::Entity_kind::CELL, AmanziMesh::Parallel_kind::OWNED);
+
+  // SW map from pipe -> SW domain
+  if (pipe_flag_ == true) {
+    pipe_map_.resize(ncells_pipe);
+    for (int c_pipe = 0; c_pipe < ncells_pipe; ++c_pipe) {
+      if (std::abs(mnhMask[0][c_pipe] - 1.0) < 1.e-12) {
+        const Amanzi::AmanziGeometry::Point& xc_pipe = pipe_mesh->getCellCentroid(c_pipe);
+        for (int c_sw = 0; c_sw < ncells_sw; ++c_sw) {
+          std::vector<AmanziGeometry::Point> coords;
+          auto cnodes = surface_mesh->getCellNodes(c_sw);
+          for (int node_sw = 0; node_sw < cnodes.size(); node_sw++) {
+            coords.push_back(surface_mesh->getNodeCoordinate(node_sw));
+          }
+
+          if (AmanziGeometry::point_in_polygon(xc_pipe, coords) == true) {
+            pipe_map_[c_pipe] = c_sw;
+            break;
+          }
+        }
+      }
+    }
+    pipe_map_created_ = true;
+  }
+
+  // Pipe map from SW -> pipe domain
+  if (sw_flag_ == true) {
+    sw_map_.resize(ncells_sw);
+    for (int c_sw = 0; c_sw < ncells_sw; ++c_sw) {
+      if (std::abs(mnhMask[0][c_sw] - 1.0) < 1.e-12) {
+        const Amanzi::AmanziGeometry::Point& xc_sw = surface_mesh->getCellCentroid(c_sw);
+        for (int c_pipe = 0; c_pipe < ncells_pipe; ++c_pipe) {
+          std::vector<AmanziGeometry::Point> coords;
+          auto cnodes = pipe_mesh->getCellNodes(c_pipe);
+          for (int node_pipe = 0; node_pipe < cnodes.size(); node_pipe++) {
+            coords.push_back(pipe_mesh->getNodeCoordinate(node_pipe));
+          }
+
+          if (AmanziGeometry::point_in_polygon(xc_sw, coords) == true) {
+            sw_map_[c_sw] = c_pipe;
+            break;
+          }
+        }
+      }
+    }
+    sw_map_created_ = true;
+  }
+}
+
+void
+PipeDrainEvaluator::EnsureCompatibility_ToDeps_(State& S, const CompositeVectorSpace& fac)
+{
+  auto domain1 = Keys::getDomain(my_keys_.front().first);
+  if (domain1 == pipe_domain_name_) {
+    for (const auto& dep : dependencies_) {
+      auto domain = Keys::getDomain(dep.first);
+      if (pipe_domain_name_ == domain) {
```

Comment:
What happens if the domain isn't pipe_domain_name?  You should still set this evaluator's expectations of what the dependency's structure is.

Likely you should be making two "fac" variables, one with pipe_domain, one with surface_domain as the mesh.  Then loop over each dependency and Update with the correct one.

Reply:
@dasvyat could you please help with this comment?